cmake/modules/OmrHookgen.cmake

```diff
@@ -23,21 +23,61 @@ if(OMR_HOOKGEN_)
 	return()
 endif()
 set(OMR_HOOKGEN_ 1)
+include(CMakeParseArguments)
 
 # Process an input hookgen file to generate output source files.
-# Usage: omr_add_hookgen(<input> <output>...)
-# TODO: currently output in source tree, should be in build tree
-# TODO: Dependecy checking is broken, since it checks for output in build tree rather than src
-function(omr_add_hookgen input)
-	get_filename_component(input_dir "${input}" DIRECTORY)
+# Usage: omr_add_hookgen(INPUT <input> [PRIVATE_DIR <private_header_dir>] [PUBLIC_DIR <public_header_dir>]
+# public headers are written to public_header_dir (defaults to current binary dir)
+# private headers are written to private_heade_dir (defautls to current binary dir)
+function(omr_add_hookgen)
+	cmake_parse_arguments(OPT "" "INPUT;PRIVATE_DIR;PUBLIC_DIR" "TARGETS" ${ARGN})
+
+	if(NOT OPT_INPUT)
+		message(FATAL_ERROR "value for INPUT must be specified")
+	endif()
+
+	#if private_dir or public_dir not specified, default to CMAKE_CURRENT_BINARY_DIR
+	if(NOT OPT_PRIVATE_DIR)
+		set(OPT_PRIVATE_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+	endif()
+	if(NOT OPT_PUBLIC_DIR)
+		set(OPT_PUBLIC_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+	endif()
+
+	get_filename_component(input_filename "${OPT_INPUT}" NAME)
+
+	file(READ "${OPT_INPUT}" hook_def)
+	string(REGEX MATCH "<privateHeader>[^<]*" private_header "${hook_def}")
```

Comment:
I think this would be clearer (and shorter) as: 
```cmake
string(REGEX REPLACE "<privateHeader>(.*?)</privateHeader>" "\\1" private_header "${hook_def}")
```